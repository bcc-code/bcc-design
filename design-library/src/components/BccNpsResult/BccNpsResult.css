@layer components {
  .bcc-nps-result {
    @apply relative w-full;
  }

  .bcc-nps-result.lg {
    width: 250px;
  }
  .bcc-nps-result.md {
    width: 177px;
  }
  .bcc-nps-result.sm {
    width: 110px;
  }
  .bcc-nps-result.tiny {
    width: 40px;
  }

  .bcc-nps-result .result-gauge-dial {
    transform-origin: center;
  }

  .bcc-nps-result.animated .result-gauge-dial {
    transition-delay: 0.3s;
    transition: all 0.7s cubic-bezier(0.75, -0.05, 0.62, 1.6);
  }

  .bcc-nps-result-labels {
    @apply absolute bottom-0 flex w-full flex-col items-center justify-center;
  }
}
