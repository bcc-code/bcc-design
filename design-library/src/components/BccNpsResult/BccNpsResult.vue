<script setup lang="ts">
import { computed, ref } from "vue";

type Props = {
  score: number; // Between -100 and 100
  size?: "lg" | "md" | "sm" | "xs";
  display?: string;
  underline?: string;
  hideText?: boolean;
  animated?: boolean;
};

const props = withDefaults(defineProps<Props>(), {
  size: "md",
  hideText: false,
  animated: true,
});

const TOT_DEGREES = 250; // 70% of 360
const MIN_DEGREES = -170;
const MAX_DEGREES = 80;

function scoreToDegrees(score: number) {
  const perc = (score + 100) / 200;
  const degr = perc * TOT_DEGREES + MIN_DEGREES;
  if (degr > MAX_DEGREES) return MAX_DEGREES;
  if (degr < MIN_DEGREES) return MIN_DEGREES;
  return degr;
}

const started = ref(false);
setTimeout(() => {
  started.value = true;
}, 150);

const degrees = computed(() => {
  const startingPoint = MIN_DEGREES;

  if (started.value && props.score !== undefined) {
    return scoreToDegrees(props.score);
  }
  return startingPoint;
});
</script>

<template>
  <div class="bcc-nps-result" :class="{ animated }">
    <svg class="result-gauge" viewBox="0 0 180 180" fill="none" xmlns="http://www.w3.org/2000/svg">
      <path
        d="M40.7001 85.7001C39.7001 97.3001 42.8001 108.9 49.5001 118.4L31.1001 131.3L16.4001 141.6C4.10011 124.3 -1.49989 103.2 0.300108 82.2001C2.10011 61.5001 11.1001 42.0001 25.6001 27.1001L38.3001 39.8001L54.2001 55.7001C46.4001 63.8001 41.6001 74.4001 40.6001 85.7001H40.7001Z"
        fill="#FFE3E3"
      />
      <path
        d="M89.0001 0V40.5C76.6001 40.7 64.7001 45.7 55.7001 54.3L39.8001 38.4L27.1001 25.7C43.6001 9.5 65.8001 0.3 89.0001 0Z"
        fill="#FFF5ED"
      />
      <path
        d="M152.9 25.7L140.2 38.4L124.3 54.3C115.3 45.7 103.5 40.8 91 40.5V0C114.2 0.2 136.4 9.4 152.9 25.6V25.7Z"
        fill="#FEFDE8"
      />
      <path
        d="M180 89.0001H139.5C139.3 76.6001 134.3 64.7001 125.7 55.7001L141.6 39.8001L154.3 27.1001C170.5 43.7001 179.7 65.9001 179.9 89.0001H180Z"
        fill="#D7FFE7"
      />
      <path
        d="M180 91C179.8 110.1 173.5 128.7 162 144L147.6 133.2L129.6 119.7C135.8 111.4 139.3 101.4 139.5 91H180Z"
        fill="#B2FFD1"
      />
      <path
        d="M31 131.3L16.3 141.6C4.10005 124.3 -1.49995 103.2 0.300047 82.2001C2.10005 61.5001 11.1 42.0001 25.6 27.1001L38.3 39.8001C26.7 51.7001 19.6 67.2001 18.2 83.7001C16.7 100.6 21.2 117.4 31 131.3Z"
        fill="#F93A4F"
      />
      <path
        d="M89.0001 0V18C70.6001 18.2 53.0001 25.5 39.8001 38.4L27.1001 25.7C43.6001 9.5 65.8001 0.3 89.0001 0Z"
        fill="#FE7F39"
      />
      <path
        d="M152.9 25.7L140.2 38.4C127 25.6 109.4 18.3 91 18V0C114.2 0.2 136.4 9.4 152.9 25.6V25.7Z"
        fill="#FFEE45"
      />
      <path
        d="M180 89.0001H162C161.8 70.6001 154.5 53.0001 141.6 39.8001L154.3 27.1001C170.5 43.7001 179.7 65.9001 179.9 89.0001H180Z"
        fill="#7DFFB3"
      />
      <path
        d="M180 91C179.8 110.1 173.5 128.7 162 144L147.6 133.2C156.8 121 161.8 106.2 162 91H180Z"
        fill="#09DE62"
      />

      <path
        d="M128.292 49.4565L92.4146 83.7227C91.5428 83.4108 90.5855 83.2892 89.604 83.4082C86.2246 83.8177 83.816 86.8941 84.226 90.2772C84.636 93.6603 87.7097 96.0722 91.089 95.6627C94.4684 95.2532 96.877 92.1769 96.4671 88.7937C96.4262 88.4568 96.3578 88.1298 96.2663 87.8144L130.221 51.3721C131.359 50.0885 129.568 48.3071 128.292 49.4565Z"
        fill="black"
        class="result-gauge-dial"
        :style="`transform: rotate(${degrees}deg)`"
      />
    </svg>
    <div v-if="!hideText && size !== 'xs'" class="bcc-nps-result-labels" :class="size">
      <h3 class="bcc-nps-result-labels--heading">{{ display ?? score }}</h3>
      <label class="bcc-nps-result-labels--label">{{ underline }}</label>
    </div>
  </div>
</template>
