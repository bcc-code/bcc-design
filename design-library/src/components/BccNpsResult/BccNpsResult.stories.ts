import { ref } from "vue";
import BccNpsResult from "./BccNpsResult.vue";

import type { Meta, StoryFn } from "@storybook/vue3";

export default {
  title: "Feedback/BccNpsResult",
  component: BccNpsResult,
  argTypes: {
    size: {
      description: "Size of the NPS Result gauge",
      options: ["lg", "md", "sm", "tiny"],
      control: { type: "radio" },
    },
  },
} as Meta<typeof BccNpsResult>;

const Template: StoryFn<typeof BccNpsResult> = (args) => ({
  components: { BccNpsResult },
  setup() {
    const value = ref(0);
    return { args, value };
  },
  template: `
    <BccNpsResult v-bind="args" v-model="value" />
  `,
});

export const Example = Template.bind({});
Example.args = { score: 30, underline: "NPS Score" };
