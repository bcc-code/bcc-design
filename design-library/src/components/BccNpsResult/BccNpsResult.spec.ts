import { describe, it, expect } from "vitest";
import { mount } from "@vue/test-utils";
import BccNpsResult from "./BccNpsResult.vue";

describe("BccNpsResult", () => {
  it("renders a nps result", () => {
    expect(BccNpsResult).toBeTruthy();

    const wrapper = mount(BccNpsResult, {
      props: { score: 20 },
    });

    expect(wrapper.html()).toMatchSnapshot();
  });
});
