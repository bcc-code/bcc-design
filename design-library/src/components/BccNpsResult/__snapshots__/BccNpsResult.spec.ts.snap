// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BccNpsResult > renders a nps result 1`] = `
"<div class=\\"bcc-nps-result animated\\"><svg class=\\"result-gauge\\" viewBox=\\"0 0 180 180\\" fill=\\"none\\" xmlns=\\"http://www.w3.org/2000/svg\\">
    <path d=\\"M16.2765 141.622C4.14125 124.291 -1.50042 103.239 0.342066 82.1622C2.18456 61.0855 11.3934 41.3317 26.3516 26.3693L54.9935 55.0031C46.7664 63.2324 41.7016 74.097 40.6882 85.6892C39.6748 97.2814 42.7777 108.86 49.4521 118.392L16.2765 141.622Z\\" fill=\\"#FFE3E3\\"></path>
    <path d=\\"M26.3605 26.3604C43.2388 9.48211 66.1307 -2.84641e-07 90.0001 0V40.5C76.8719 40.5 64.2814 45.7152 54.9984 54.9982L26.3605 26.3604Z\\" fill=\\"#FFE9D4\\"></path>
    <path d=\\"M90.0001 0C113.87 2.84641e-07 136.761 9.48211 153.64 26.3604L125.002 54.9982C115.719 45.7152 103.128 40.5 90.0001 40.5V0Z\\" fill=\\"#FEFDE8\\"></path>
    <path d=\\"M153.64 26.3604C170.518 43.2387 180 66.1305 180 90L139.5 90C139.5 76.8718 134.285 64.2813 125.002 54.9982L153.64 26.3604Z\\" fill=\\"#D7FFE7\\"></path>
    <path d=\\"M180 90C180 109.47 173.686 128.415 162.006 143.993L129.603 119.696C136.027 111.128 139.5 100.709 139.5 90L180 90Z\\" fill=\\"#DCEEE4\\"></path>
    <path d=\\"M16.2765 141.622C4.14125 124.291 -1.50042 103.239 0.342066 82.1622C2.18456 61.0855 11.3934 41.3317 26.3516 26.3693L39.0813 39.0954C27.1148 51.0653 19.7477 66.8684 18.2737 83.7298C16.7997 100.591 21.313 117.433 31.0212 131.298L16.2765 141.622Z\\" fill=\\"#F93A4F\\"></path>
    <path d=\\"M26.3605 26.3604C43.2388 9.48211 66.1307 -2.84641e-07 90.0001 0V18C70.9046 18 52.5911 25.5857 39.0885 39.0883L26.3605 26.3604Z\\" fill=\\"#FE7F39\\"></path>
    <path d=\\"M90.0001 0C113.87 2.84641e-07 136.761 9.48211 153.64 26.3604L140.912 39.0883C127.409 25.5857 109.096 18 90.0001 18V0Z\\" fill=\\"#FCDD13\\"></path>
    <path d=\\"M153.64 26.3604C170.518 43.2387 180 66.1305 180 90H162C162 70.9044 154.414 52.5909 140.912 39.0883L153.64 26.3604Z\\" fill=\\"#09DE62\\"></path>
    <path d=\\"M180 90C180 109.47 173.686 128.415 162.006 143.993L147.605 133.194C156.949 120.732 162 105.576 162 90H180Z\\" fill=\\"#01B84E\\"></path>
    <path d=\\"M128.292 49.4565L92.4146 83.7227C91.5428 83.4108 90.5855 83.2892 89.604 83.4082C86.2246 83.8177 83.816 86.8941 84.226 90.2772C84.636 93.6603 87.7097 96.0722 91.089 95.6627C94.4684 95.2532 96.877 92.1769 96.4671 88.7937C96.4262 88.4568 96.3578 88.1298 96.2663 87.8144L130.221 51.3721C131.359 50.0885 129.568 48.3071 128.292 49.4565Z\\" fill=\\"black\\" class=\\"result-gauge-dial\\" style=\\"transform: rotate(-170deg);\\"></path>
  </svg>
  <div class=\\"bcc-nps-result-labels md\\">
    <h3 class=\\"bcc-nps-result-labels--heading\\">20</h3><label class=\\"bcc-nps-result-labels--label\\"></label>
  </div>
</div>"
`;
